server/metrics/metrics_service.ts

```diff
@@ -0,0 +1,152 @@
+import { METRIC_INTERVAL, DEFAULT_WINDOW_SIZE } from '.';
+
+interface MetricValue {
+  sum: number;
+  count: number;
+}
+
+interface MetricOutput {
+  response_time_avg: number;
+  requests_per_second: number;
+}
+
+interface MetricsData {
+  [componentName: string]: {
+    [action: string]: {
+      [statusCode: number]: MetricValue;
+    };
+  };
+}
+
+export interface Stats {
+  readonly data: MetricsData;
+  readonly overall: MetricOutput;
+  readonly component_counts: Record<string, number>;
+  readonly status_code_counts: Record<string, number>;
+}
+
+export interface MetricsServiceSetup {
+  addMetric: (componentName: string, action: string, statusCode: number, value: number) => void;
+  getStats: () => Stats;
+}
+
+export class MetricsService {
+  private interval: number = METRIC_INTERVAL.ONE_MINUTE;
+  private windowSize: number = DEFAULT_WINDOW_SIZE;
+
+  private data: Record<number, MetricsData> = {};
+  private componentCounts: Record<number, Record<string, number>> = {};
+  private statusCodeCounts: Record<number, Record<string, number>> = {};
+  private overall: Record<number, MetricValue> = {};
+
+  constructor(private logger?: Logger) {}
+
+  setup(
+    interval: number = METRIC_INTERVAL.ONE_MINUTE,
+    windowSize: number = DEFAULT_WINDOW_SIZE
+  ): MetricsServiceSetup {
+    this.interval = interval;
+    this.windowSize = windowSize;
+
+    const addMetric = (
+      componentName: string,
+      action: string,
+      statusCode: number,
+      value: number
+    ): void => {
+      const currInterval = Math.floor(Date.now() / this.interval);
+
+      this.trim();
+
+      if (!this.data[currInterval]) {
+        this.data[currInterval] = {};
+        this.overall[currInterval] = { sum: 0, count: 0 };
+        this.componentCounts[currInterval] = {};
+        this.statusCodeCounts[currInterval] = {};
+      }
+
+      if (!this.data[currInterval][componentName]) {
+        this.data[currInterval][componentName] = {};
+        this.componentCounts[currInterval][componentName] = 0;
+      }
+
+      if (!this.data[currInterval][componentName][action]) {
+        this.data[currInterval][componentName][action] = {};
+      }
+
+      if (!this.data[currInterval][componentName][action][statusCode]) {
+        this.data[currInterval][componentName][action][statusCode] = { sum: 0, count: 0 };
+        this.statusCodeCounts[currInterval][statusCode] = 0;
+      }
+
+      const { sum, count } = this.data[currInterval][componentName][action][statusCode];
+      this.data[currInterval][componentName][action][statusCode] = {
+        sum: sum + value,
+        count: count + 1,
+      };
+
+      this.componentCounts[currInterval][componentName]++;
+      this.statusCodeCounts[currInterval][statusCode]++;
+
+      this.overall[currInterval].sum += value;
+      this.overall[currInterval].count++;
+    };
+
+    const getStats = (): Stats => {
+      const prevInterval = Math.floor(Date.now() / this.interval) - 1;
```

Comment:
This is the entire minute prior to the current minute, right?

So, let's say that I call `getStats` at 6:15:05, then it will return all stats added between 6:14:00.000 and 6:14:59.999 (inclusive), right?

Reply:
correct